File: server/internal/plansig/plansig.go
// Package plansig signs install plans with the backend's long-lived
// Ed25519 key, so a plan fetched over plain HTTP (old machines, design.md
// 1.3) can't be swapped or edited on the way (docs/format.md section 3.2).
//
// A signed plan is the plan's exact bytes followed by one last line:
//
//	sig<TAB>ed25519<TAB><base64 of the 64-byte signature>\n
//
// The signature (RFC 8032 Ed25519, no prehash, no context) covers every
// byte before that line, starting with the `ib-plan` header and ending
// with the newline just before `sig`. The plan's `record` line is inside
// the signed bytes, which is what binds a plan to one record: engines
// refuse a plan whose `record` isn't the record they are installing.
package plansig

import (
	"bytes"
	"crypto/ed25519"
	"crypto/rand"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/hex"
	"encoding/pem"
	"errors"
	"fmt"
	"io/fs"
	"log"
	"os"
	"path/filepath"
	"strings"
)

// File names in the data folder.
const (
	KeyFile = "plan-signing-key.pem" // PKCS#8 private key, mode 0600
	PubFile = "plan-signing-key.pub" // base64 of the raw 32-byte public key; bases are built with it
)

const sigPrefix = "sig\ted25519\t"

// Signer holds the key.
type Signer struct {
	priv ed25519.PrivateKey
	Pub  ed25519.PublicKey
}

// LoadOrCreate reads the key from dir, or makes one if there is none. The
// public key file is (re)written from the private key every time, so it
// can't drift from it.
func LoadOrCreate(dir string) (s *Signer, created bool, err error) {
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return nil, false, err
	}
	kp := filepath.Join(dir, KeyFile)
	b, err := os.ReadFile(kp)
	switch {
	case errors.Is(err, fs.ErrNotExist):
		_, priv, gerr := ed25519.GenerateKey(rand.Reader)
		if gerr != nil {
			return nil, false, gerr
		}
		der, merr := x509.MarshalPKCS8PrivateKey(priv)
		if merr != nil {
			return nil, false, merr
		}
		f, oerr := os.OpenFile(kp, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0o600)
		if oerr != nil {
			// Another process made it first: use theirs.
			if errors.Is(oerr, fs.ErrExist) {
				return LoadOrCreate(dir)
			}
			return nil, false, oerr
		}
		werr := pem.Encode(f, &pem.Block{Type: "PRIVATE KEY", Bytes: der})
		if cerr := f.Close(); werr == nil {
			werr = cerr
		}
		if werr != nil {
			os.Remove(kp)
			return nil, false, werr
		}
		s, created = &Signer{priv: priv, Pub: priv.Public().(ed25519.PublicKey)}, true
	case err != nil:
		return nil, false, err
	default:
		if st, serr := os.Stat(kp); serr == nil && st.Mode().Perm()&0o077 != 0 {
			log.Printf("plansig: WARNING: %s is readable by other users (mode %v); chmod 600 it", kp, st.Mode().Perm())
		}
		blk, _ := pem.Decode(b)
		if blk == nil || blk.Type != "PRIVATE KEY" {
			return nil, false, fmt.Errorf("%s: not a PEM private key", kp)
		}
		k, perr := x509.ParsePKCS8PrivateKey(blk.Bytes)
		if perr != nil {
			return nil, false, fmt.Errorf("%s: %w", kp, perr)
		}
		priv, ok := k.(ed25519.PrivateKey)
		if !ok {
			return nil, false, fmt.Errorf("%s: not an Ed25519 key", kp)
		}
		s = &Signer{priv: priv, Pub: priv.Public().(ed25519.PublicKey)}
	}
	pub := []byte(s.PublicBase64() + "\n")
	pp := filepath.Join(dir, PubFile)
	if old, _ := os.ReadFile(pp); !bytes.Equal(old, pub) {
		if err := os.WriteFile(pp+".tmp", pub, 0o644); err != nil {
			return nil, false, err
		}
		if err := os.Rename(pp+".tmp", pp); err != nil {
			return nil, false, err
		}
	}
	return s, created, nil
}

// PublicBase64 is the raw 32-byte public key in standard base64 (44 characters).
func (s *Signer) PublicBase64() string { return base64.StdEncoding.EncodeToString(s.Pub) }

// PublicPEM is the public key as a SubjectPublicKeyInfo PEM (what openssl reads).
func (s *Signer) PublicPEM() string {
	der, _ := x509.MarshalPKIXPublicKey(s.Pub)
	return string(pem.EncodeToMemory(&pem.Block{Type: "PUBLIC KEY", Bytes: der}))
}

// KeyID is a short fingerprint for logs and error messages: the first 16
// hex digits of the SHA-256 of the raw public key.
func KeyID(pub ed25519.PublicKey) string {
	h := sha256.Sum256(pub)
	return hex.EncodeToString(h[:8])
}

// Sign returns plan with its signature line appended. The plan must be an
// ib-plan and not already signed; a missing final newline is added first
// (it is part of the signed bytes).
func (s *Signer) Sign(plan []byte) ([]byte, error) {
	if !bytes.HasPrefix(plan, []byte("ib-plan\t")) {
		return nil, errors.New("plansig: not an ib-plan")
	}
	if _, _, signed := Split(plan); signed {
		return nil, errors.New("plansig: already signed")
	}
	out := make([]byte, 0, len(plan)+1+len(sigPrefix)+89)
	out = append(out, plan...)
	if len(out) > 0 && out[len(out)-1] != '\n' {
		out = append(out, '\n')
	}
	sig := ed25519.Sign(s.priv, out)
	out = append(out, sigPrefix...)
	out = append(out, base64.StdEncoding.EncodeToString(sig)...)
	out = append(out, '\n')
	return out, nil
}

// SignString is Sign for strings.
func (s *Signer) SignString(plan string) (string, error) {
	b, err := s.Sign([]byte(plan))
	return string(b), err
}

// Split finds the signature line. ok is false when the last line isn't a
// `sig` line (an unsigned plan). The last line is what follows the last
// newline, once one final "\n" (or "\r\n") is set aside.
func Split(doc []byte) (msg []byte, sigLine string, ok bool) {
	end := len(doc)
	if end > 0 && doc[end-1] == '\n' {
		end--
		if end > 0 && doc[end-1] == '\r' {
			end--
		}
	}
	nl := bytes.LastIndexByte(doc[:end], '\n')
	if nl < 0 {
		return doc, "", false
	}
	last := string(doc[nl+1 : end])
	if last != "sig" && !strings.HasPrefix(last, "sig\t") {
		return doc, "", false
	}
	return doc[:nl+1], last, true
}

// Errors from Verify.
var (
	ErrUnsigned = errors.New("the plan is not signed")
	ErrBadSig   = errors.New("the plan's signature does not verify")
)

// Verify checks a signed plan against pub and returns the signed bytes
// (the plan without its signature line).
func Verify(pub ed25519.PublicKey, doc []byte) ([]byte, error) {
	msg, line, ok := Split(doc)
	if !ok {
		return nil, ErrUnsigned
	}
	if !strings.HasPrefix(line, sigPrefix) {
		return nil, fmt.Errorf("%w: unknown signature type %q", ErrBadSig, line)
	}
	b64 := line[len(sigPrefix):]
	sig, err := base64.StdEncoding.Strict().DecodeString(b64)
	if err != nil || len(sig) != ed25519.SignatureSize || len(b64) != 88 {
		return nil, fmt.Errorf("%w: malformed signature", ErrBadSig)
	}
	if !bytes.HasPrefix(msg, []byte("ib-plan\t")) {
		return nil, fmt.Errorf("%w: signed bytes are not an ib-plan", ErrBadSig)
	}
	if !ed25519.Verify(pub, msg, sig) {
		return nil, ErrBadSig
	}
	return msg, nil
}

// VerifyFor is Verify plus the record binding: the plan's header `record`
// line must be exactly record.
func VerifyFor(pub ed25519.PublicKey, doc []byte, record string) ([]byte, error) {
	msg, err := Verify(pub, doc)
	if err != nil {
		return nil, err
	}
	if got := RecordOf(msg); got != record {
		return nil, fmt.Errorf("the plan is for record %q, not %q", got, record)
	}
	return msg, nil
}

// RecordOf returns the value of the plan header's first `record` line.
func RecordOf(plan []byte) string {
	for _, raw := range strings.Split(string(plan), "\n") {
		raw = strings.TrimSuffix(raw, "\r")
		if raw == "[target]" {
			break
		}
		if v, ok := strings.CutPrefix(raw, "record\t"); ok {
			v, _, _ = strings.Cut(v, "\t")
			return v
		}
	}
	return ""
}

File: src/build_server/lib/plansig.js
// Plan signatures (docs/format.md "Plan signature"; the Go server's plansig
// package), with node:crypto. A signed plan is the plan's exact bytes and
// one last line:
//
//   sig<TAB>ed25519<TAB><base64 of the 64-byte signature>\n
//
// The signature (RFC 8032 Ed25519, no prehash, no context) covers every
// byte before that line. Ed25519 is deterministic, so the same key and bytes
// give the same signature as the Go server.
import crypto from 'node:crypto';
import fs from 'node:fs';
import path from 'node:path';

export const KEY_FILE = 'plan-signing-key.pem'; // PKCS#8 private key, mode 0600
export const PUB_FILE = 'plan-signing-key.pub'; // base64 of the raw 32-byte public key; bases are built with it

const SIG_PREFIX = 'sig\ted25519\t';
const PLAN_HEAD = Buffer.from('ti-plan\t');
// The other document signed with the plan key: the revocation list
// (design.md 7.1). Its signed bytes must start with its own header, so a
// plan signature can never be read as one, or the other way round.
export const PLAN_KIND = 'ti-plan';
export const REVOCATIONS_KIND = 'ti-revocations';
export const RELEASES_KIND = 'ti-releases';
const headOf = (kind) => Buffer.from(kind + '\t');

function rawPublic(keyObject) {
  return Buffer.from(keyObject.export({ format: 'jwk' }).x, 'base64url');
}

export function publicKeyFromRaw(raw) {
  return crypto.createPublicKey({ key: { kty: 'OKP', crv: 'Ed25519', x: Buffer.from(raw).toString('base64url') }, format: 'jwk' });
}

// KeyID: the first 16 hex digits of the SHA-256 of the raw public key.
export function keyID(raw) {
  return crypto.createHash('sha256').update(raw).digest('hex').slice(0, 16);
}

export class Signer {
  constructor(privateKey) {
    if (privateKey.asymmetricKeyType !== 'ed25519') throw new Error('not an Ed25519 key');
    this.priv = privateKey;
    this.pubKey = crypto.createPublicKey(privateKey);
    this.pub = rawPublic(this.pubKey);
  }

  publicBase64() { return this.pub.toString('base64'); }

  // SubjectPublicKeyInfo PEM, what openssl reads.
  publicPEM() { return this.pubKey.export({ type: 'spki', format: 'pem' }); }

  // plan (bytes or text) with its signature line appended. It must be an
  // ti-plan and not already signed; a missing final newline is added first.
  sign(plan) { return this.signAs(PLAN_KIND, plan); }

  // The same for another document signed with this key: its bytes must
  // start with `<kind><TAB>`.
  signAs(kind, doc) {
    let b = Buffer.from(doc);
    const head = headOf(kind);
    if (!b.subarray(0, head.length).equals(head)) throw new Error('plansig: not a ' + kind);
    if (split(b).ok) throw new Error('plansig: already signed');
    if (b.length && b[b.length - 1] !== 0x0a) b = Buffer.concat([b, Buffer.from('\n')]);
    const sig = crypto.sign(null, b, this.priv);
    return Buffer.concat([b, Buffer.from(SIG_PREFIX + sig.toString('base64') + '\n')]);
  }

  signString(plan) { return this.sign(Buffer.from(plan, 'utf8')).toString('utf8'); }
  signStringAs(kind, doc) { return this.signAs(kind, Buffer.from(doc, 'utf8')).toString('utf8'); }
}

// Where the private key lives, when nothing says otherwise.
//
// Not the data directory, and not anywhere under the repository. The
// repository is going public and the thing that must never leave is the
// one file that cannot be replaced: every installer ever built carries
// the public half, so losing this key ends the chain for all of them and
// leaking it lets somebody else sign as us. Keeping it out of the tree
// means no .gitignore has to be right for it to stay private.
//
// $TI_KEYS overrides, and -keys on the server.
export function defaultKeyDir() {
  return process.env.TI_KEYS ||
    path.join(process.env.HOME || process.env.USERPROFILE || '.', '.config', 'tiddlyinstall', 'keys');
}

// loadOrCreate reads the key from dir, or makes one if there is none. The
// public key file is (re)written from the private key every time, so it
// can't drift from it. `alsoPub` gets a copy of the public half: the base
// builds read it from the data directory, and it is not a secret.
// Returns {signer, created}.
export function loadOrCreate(dir, log = console.log, alsoPub = '') {
  fs.mkdirSync(dir, { recursive: true, mode: 0o700 });
  const kp = path.join(dir, KEY_FILE);
  let signer, created = false;
  let pem = null;
  try { pem = fs.readFileSync(kp, 'utf8'); } catch (e) { if (e.code !== 'ENOENT') throw e; }
  if (pem === null) {
    const { privateKey } = crypto.generateKeyPairSync('ed25519');
    const out = privateKey.export({ type: 'pkcs8', format: 'pem' });
    try {
      fs.writeFileSync(kp, out, { mode: 0o600, flag: 'wx' });
    } catch (e) {
      if (e.code === 'EEXIST') return loadOrCreate(dir, log); // another process made it first: use theirs
      throw e;
    }
    signer = new Signer(privateKey);
    created = true;
  } else {
    const st = fs.statSync(kp);
    if (st.mode & 0o077) log(`plansig: WARNING: ${kp} is readable by other users (mode ${(st.mode & 0o777).toString(8)}); chmod 600 it`);
    if (!/^-----BEGIN PRIVATE KEY-----/m.test(pem)) throw new Error(kp + ': not a PEM private key');
    let key;
    try { key = crypto.createPrivateKey({ key: pem, format: 'pem' }); } catch (e) { throw new Error(kp + ': ' + e.message); }
    if (key.asymmetricKeyType !== 'ed25519') throw new Error(kp + ': not an Ed25519 key');
    signer = new Signer(key);
  }
  const pub = signer.publicBase64() + '\n';
  for (const d of alsoPub && alsoPub !== dir ? [dir, alsoPub] : [dir]) {
    const pp = path.join(d, PUB_FILE);
    let old = null;
    try { old = fs.readFileSync(pp, 'utf8'); } catch (e) { /* none yet */ }
    if (old !== pub) {
      fs.mkdirSync(d, { recursive: true });
      fs.writeFileSync(pp + '.tmp', pub, { mode: 0o644 });
      fs.renameSync(pp + '.tmp', pp);
    }
  }
  return { signer, created };
}

// split finds the signature line: {msg, line, ok}. ok is false when the last
// line (what follows the last newline, once one final "\n" or "\r\n" is set
// aside) isn't a `sig` line.
export function split(doc) {
  const b = Buffer.from(doc);
  let end = b.length;
  if (end > 0 && b[end - 1] === 0x0a) {
    end--;
    if (end > 0 && b[end - 1] === 0x0d) end--;
  }
  const nl = b.subarray(0, end).lastIndexOf(0x0a);
  if (nl < 0) return { msg: b, line: '', ok: false };
  const last = b.subarray(nl + 1, end).toString('utf8');
  if (last !== 'sig' && !last.startsWith('sig\t')) return { msg: b, line: '', ok: false };
  return { msg: b.subarray(0, nl + 1), line: last, ok: true };
}

export class VerifyError extends Error {}

// verify checks a signed plan against the raw public key and returns the
// signed bytes (the plan without its signature line). `kind` is the header
// the signed bytes must start with: a ti-plan unless another is named.
export function verify(pubRaw, doc, kind = PLAN_KIND) {
  const { msg, line, ok } = split(doc);
  if (!ok) throw new VerifyError('the plan is not signed');
  if (!line.startsWith(SIG_PREFIX)) throw new VerifyError('the plan\'s signature does not verify: unknown signature type ' + JSON.stringify(line));
  const b64 = line.slice(SIG_PREFIX.length);
  const sig = Buffer.from(b64, 'base64');
  if (b64.length !== 88 || !/^[A-Za-z0-9+/]{86}==$/.test(b64) || sig.length !== 64) throw new VerifyError('the plan\'s signature does not verify: malformed signature');
  const head = headOf(kind);
  if (!msg.subarray(0, head.length).equals(head)) throw new VerifyError('the plan\'s signature does not verify: signed bytes are not a ' + kind);
  if (!crypto.verify(null, msg, publicKeyFromRaw(pubRaw), sig)) throw new VerifyError('the plan\'s signature does not verify');
  return msg;
}

// verifyFor is verify plus the record binding: the plan's header `record`
// line must be exactly `record`.
export function verifyFor(pubRaw, doc, record) {
  const msg = verify(pubRaw, doc);
  const got = recordOf(msg);
  if (got !== record) throw new VerifyError(`the plan is for record "${got}", not "${record}"`);
  return msg;
}

// recordOf: the value of the plan header's first `record` line.
export function recordOf(plan) {
  for (let raw of Buffer.from(plan).toString('utf8').split('\n')) {
    raw = raw.replace(/\r$/, '');
    if (raw === '[target]') break;
    if (raw.startsWith('record\t')) return raw.slice(7).split('\t')[0];
  }
  return '';
}

// AddRequestLine puts `request<TAB>vals...` after the plan's header line and
// after any `request` lines already there, so the lines keep the order they
// were added in: `request<TAB>name<TAB>…` (a plan by name, format.md section
// 2) before `request<TAB>nonce<TAB>…` (design.md 7.1). The order is part of
// the format: an engine that reads one `request` line reads the first, which
// is the one saying what was asked for.
export function addRequestLine(plan, ...vals) {
  let i = plan.indexOf('\n');
  if (i < 0 || !plan.startsWith('ti-plan\t')) throw new Error('not a ti-plan');
  for (;;) {
    if (!plan.startsWith('request\t', i + 1)) break;
    const n = plan.indexOf('\n', i + 1);
    if (n < 0) break;
    i = n;
  }
  const line = ['request', ...vals.map((v) => String(v).replace(/[\t\r\n]/g, ' '))].join('\t') + '\n';
  return plan.slice(0, i + 1) + line + plan.slice(i + 1);
}
